aten/src/ATen/native/IndexingUtils.cpp

```diff
@@ -0,0 +1,32 @@
+
+#include <ATen/native/IndexingUtils.h>
+
+namespace at { namespace native {
+
+bool canUse32BitIndexMath(const Tensor& t, int64_t max_elem) {
+  int64_t elements = t.numel();
+  if (elements >= max_elem) {
+    return false;
+  }
+  if (elements == 0) {
+    return max_elem > 0;
+  }
+
+  int64_t offset = 0;
+  int64_t linearId = elements - 1;
+
+  for (int i = t.dim() - 1; i >= 0; --i) {
+    int64_t curDimIndex = linearId % t.size(i);
+    int64_t curDimOffset = curDimIndex * t.stride(i);
```

Comment:
Slightly off topic, but I think this is wrong? I think this assumes all strides are positive. So, for example if you had a negative stride axis it will subtract from the maximum offset instead of contributing 0. And if you had entriely negative strides offset will be `< 0` and therefore not `>= max_elem` but still be out of the range of an `int`. 

I think it should be calculating something like:
```python
min, max = 0, 0
for len, stride in zip(t.size(), t.stride()):
    if stride > 0:
        max += (len - 1) * stride
    elif stride < 0:
        min += (len - 1) * stride

return min >= -max_elem and max <= max_elem
```

Does that sound right?

Reply:
Strides in PyTorch are always >= 0 so we don't have to worry about negative strides. It might be good to note here that we are assuming strides are positive just in case someone decides to implement negative strides (but there have been no serious attempts to implement negative strides in the past).

Reply:
Ah okay, I didn't realize negative strides were unsupported.